p2p/ipld/read_test.go

```diff
@@ -217,14 +223,117 @@ func TestBlockRecovery(t *testing.T) {
 				return
 			}
 
-			require.NoError(t, err)
+			assert.NoError(t, err)
 
 			// check that the squares are equal
 			assert.Equal(t, flatten(eds), flatten(reds))
 		})
 	}
 }
 
+func TestRetrieveBlockData(t *testing.T) {
+	type test struct {
+		name       string
+		squareSize int
+		expectErr  bool
+		errStr     string
+	}
+
+	// create a mock node
+	ipfsNode, err := coremock.NewMockNode()
+	if err != nil {
+		t.Error(err)
+	}
+
+	// issue a new API object
+	ipfsAPI, err := coreapi.NewCoreAPI(ipfsNode)
+	if err != nil {
+		t.Error(err)
+	}
+
+	// the max size of message that won't get splitl
+	adjustedMsgSize := types.MsgShareSize - 2
+
+	tests := []test{
+		{"4 KB block", 4, false, ""},
+		{"16 KB block", 8, false, ""},
+		{"16 KB block timeout expected", 8, true, "timeout"},
+	}
+
+	// largeTests are tests that too big to be ran with the race dectector
+	largeTests := []test{
+		{"max square size", types.MaxSquareSize, false, ""},
+	}
+
+	// if we're using the race detector, skip some large tests due to time and
+	// concurrency constraints
+	if raceDetectorActive {
+		log.Println("Not running large tests for TestRetrieveBlockData: race detector is active")
+	} else {
+		tests = append(tests, largeTests...)
+	}
```

Comment:
now we're only running the large tests when the race detector is off, because the race detector introduces constraints that cause the large tests to fail. The first constrain is that the race detector will fail if there are more than 8192 concurrent goroutines. The second constraint slows the large tests down so much as to cause them to timeout without introducing absurd timeouts.

Reply:
the race build flag, like all other build flags, are [not accessible](https://github.com/golang/go/issues/7007) to the runtime, so the hacky work around is to have an `init` func in a file that only gets built using the race flag. We are doing that in the `race_detector_file.go` file.

Reply:
Instead of splitting up the tests, I'd recommend to use `t.Skip()` if the test is larger than some threshold:
```go
if raceDetectorActive && tc.squareSize > 8 {
    t.Skip("Not running large test due to time and concurrency constraints while race detector is active.")
}
```
This yields the following output:
```
=== RUN   TestRetrieveBlockData/16_KB_block_timeout_expected_size_8
    --- PASS: TestRetrieveBlockData/16_KB_block_timeout_expected_size_8 (2.02s)
=== RUN   TestRetrieveBlockData/max_square_size_size_128
    read_test.go:268: Not running large test due to time and concurrency constraints while race detector is active.
    --- SKIP: TestRetrieveBlockData/max_square_size_size_128 (0.00s)

Test ignored.
```

Reply:
smooth :sunglasses: https://github.com/lazyledger/lazyledger-core/pull/232/commits/50b691e960531b124b95d59bcbc33d0d980a920f